clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsResponse.java

```diff
@@ -184,18 +87,75 @@ public boolean shouldClientThrottle(short version) {
 
     private void validate(short version) {
         if (version == 0) {
-            final boolean unsupported = acls.stream()
-                .map(AclBinding::pattern)
-                .map(ResourcePattern::patternType)
-                .anyMatch(patternType -> patternType != PatternType.LITERAL);
+            final boolean unsupported = acls().stream()
+                .anyMatch(acl -> acl.patternType() != PatternType.LITERAL.code());
             if (unsupported) {
                 throw new UnsupportedVersionException("Version 0 only supports literal resource pattern types");
             }
         }
 
-        final boolean unknown = acls.stream().anyMatch(AclBinding::isUnknown);
+        final boolean unknown = acls().stream()
+                .map(DescribeAclsResponse::aclBindings)
+                .anyMatch(bindings -> bindings.stream().anyMatch(AclBinding::isUnknown));
         if (unknown) {
             throw new IllegalArgumentException("Contain UNKNOWN elements");
         }
     }
+
+    private static List<AclBinding> aclBindings(DescribeAclsResource resource) {
+        List<AclBinding> acls = new ArrayList<>();
+        for (AclDescription acl : resource.acls()) {
+            ResourcePattern pattern = new ResourcePattern(
+                    ResourceType.fromCode(resource.type()),
+                    resource.name(),
+                    PatternType.fromCode(resource.patternType()));
+            AccessControlEntry entry = new AccessControlEntry(
+                    acl.principal(),
+                    acl.host(),
+                    AclOperation.fromCode(acl.operation()),
+                    AclPermissionType.fromCode(acl.permissionType()));
+            acls.add(new AclBinding(pattern, entry));
+        }
+        return acls;
+    }
+
+    public static List<AclBinding> aclBindings(List<DescribeAclsResource> resources) {
+        List<AclBinding> acls = new ArrayList<>();
+        for (DescribeAclsResource resource : resources) {
+            acls.addAll(aclBindings(resource));
+        }
+        return acls;
+    }
+
+    public static DescribeAclsResponse prepareResponse(int throttleTimeMs, ApiError error, Collection<AclBinding> acls) {
```

Comment:
Maybe this should be called `fromBindings`.

Reply:
Many responses have a `prepareResponse()` method, so I thought it made sense to call it this way for consistency. If `fromBinding()` is clearer, I'm happy to update it